### ISSUE TITLE
HLSL: Vertex/Fragment interface does not consider semantics.

### ISSUE THREAD

**Issue by Themaister** [KhronosGroup/glslang#871]
One problem right now is that semantics seem to be ignored for vertex out variables and fragment input variables, which creates a brittle interface when vertex and fragment shaders are not doing the exact same things w.r.t. I/O variables.

E.g.:
```
struct VertexOut
{
    float4 output : SEMANTIC;
};
VertexOut main() {}

void main(float4 input : SEMANTIC)
{}
```
Creates SPIR-V where vertex shader returns a struct and fragment takes a plain input variable, which is not compatible.
From what I understand, this is supposed to work in D3D.

Another problem case is when the ordering of the members of the struct is not the same.

```
struct VertexOut
{
    float4 output : SEMANTIC;
    float4 output2 : SEMANTIC2;
};
VertexOut main() {}

struct FragmentInput
{
    float4 output2 : SEMANTIC2;
    float4 output : SEMANTIC;
};
void main(FragmentInput input)
{}
```
Here, a flat struct type is generated by glslang for both input and output. But the inherent information of semantics is lost, so there is no way to make the linkage work, because the member order is now changed, but from what I understand this is still supposed to work somehow in D3D.

The only option seems to be something like flattening the structs to standalone varyings with a location, but as far as I know, user-defined mapping is only supported for attributes and resources?

**Comment by johnkslang** [comment]
Is this expected behavior specific to only some versions of DX? (We got some pretty strong indications at the beginning of this project that linkage was done by shader ordering and not by semantic.)

**Comment by Themaister** [comment]
I would like to get some clarification on that as well. I can't find any specific mention of this anywhere. The second case might be bogus, but the first case has been reported as a problem case.

If the linkage is done by ordering, and not semantics, are structs implicitly flattened, and arguments are then processed in declaration order? In that case, I think glslang would have to flatten the interface variables rather than passing structs between stages.

**Comment by johnkslang** [comment]
Glslang is indeed doing complex transforms of structures for interfaces due to the ability to mix built-in and  user-defined members in the same struct, whereas SPIR-V wants structs to be purely built-in or purely user. That might or might not relate, depending on what the actual mismatch is.

We need to start with a specific example, tied to a specific shader model, where glslang does a contradictory interface match to what is normally done, and see what the root cause of that difference is.

**Comment by johnkslang** [comment]
fxc is assigning registers in declaration order, not by semantic.  When semantics end in numbers, even if the base name (before the number) is different, it assigns "Index" to match the number, but still assigns registers in declaration order.

If you rearrange your struct, the register numbers change, to reflect declaration order.

**Comment by johnkslang** [comment]
Given that the semantics between stages *are not* effecting the register, which maps to location, I'll change the title of this to be about the struct matching the non-struct.

**Comment by johnkslang** [comment]
> Creates SPIR-V where vertex shader returns a struct and fragment takes a plain input variable, which is not compatible.

Why is this not compatible? They should each get a location 0, and the linkage is done by location.

**Comment by punkUser** [comment]
The behavior in DX is actually a fair bit more complicated... binding of the semantics between stages actually does happen based on the semantic string at runtime (when the PSO is created or similar in DX12 for instance). Indeed there is some special logic for semantic names that end in a number (so that you can create arrays of things in an interface) but ignoring that for the moment, SEMANTICA and SEMANTICB could appear in either order in different stages but they will be linked properly when the PSO gets created.

I'm not sure if it's going to be very easy to duplicate this behavior in Vulkan, depending on how much flexibility we have with the "location" parameters in these interfaces. Perhaps if we were able to use a string hash or something that would be acceptable but failing that this might end up being pretty messy...

**Comment by johnkslang** [comment]
Yes, but what we need is the rules by which we do this for Vulkan.  We define a convention, that then glslang and Vulkan apps follow, to get things linked up.  E.g., a major one is that Vulkan requires inter-stage linkage by location, glslang assigns locations in declaration order (as does fxc for registers), and it works.  We don't have to replicate DX complexities, we just have to have a convention to follow that works.

The IO remapper also helps here, as complex schemes can be implemented within it.

**Comment by punkUser** [comment]
That's what I'm saying, I don't think the Vulkan setup of static locations *does* work for arbitrary HLSL shaders. In HLSL I absolutely can have a single vertex shader with different output semantics that gets linked - properly - to several different pixel shaders that declare those inputs in different orders. That simply does not work in Vulkan right now outside of:

a) Hashing the semantic name and using that as the location or similar
b) Patching the SPIRV at link time separately for each PSO (ioremapper or similar)

B is a problematic solution as it means we can't even create the Vulkan shader module until PSO creation time, which can be quite expensive. A) could work if the values accepted for "locations" are suitably broad (i.e. arbitrary 32-bit integer or similar).

I agree we don't have to replicate everything about how DX works, but in this case we do have a potential 1:many relationship that gets late bound at pipeline linking time that it appears Vulkan doesn't have a good way to handle, unless I'm misunderstanding the spec.

**Comment by johnkslang** [comment]
Since Vulkan has no link-by name phase, it may be best to work out something offline.  Possibly using the IO remapper to implement custom algorithms, such as a hash, or Index X Register mapper, etc.

**Comment by punkUser** [comment]
Yes we'll probably have to add this to the list of things to bring up with the Vulkan API folks. In the mean time it would be good to at least get clarification on whether we can treat the "location" values as mere identifiers or if drivers are allowed to do something nasty if I set one of them to 4 billion or something ;)

**Comment by johnkslang** [comment]
It may also help to save the semantic as the OpName instead of the variable name. I think something will need to map `semantic X location` into final location, and since DX `index` comes from the number in the semantic, it should be enough information to replicate what was happening with `semantic-base X index X register` in DX.
